### DISCUSSION TITLE
Text-Input

### DISCUSSION THREAD

**Discussion by TopperDEL** [StereoKit/StereoKit#109] [Q&A] [answered]
My app would need for the user to enter text. The Mixed Reality Toolkit from Microsoft has a nice Keyboard included - even with swipe-gestures.
How would it be possible to include something like this - would you be able to add such a keyboard from them or would you (or worse: would I) need to code such a keyboard on my own?

**Comment by maluoi** [comment, marked as answer]
I do have a text input UI feature, but I'd consider it only partially implemented right now! I'd probably not rely on it until I've done another pass on it, and use Speech To Text stuff instead. It does work with physical keyboards, bluetooth/wired and all that, and if you pop the system keyboard up on your own, there's a good chance it'll work with the input already!

The plan is to have StereoKit properly bring up the system keyboard if it's available (WMR and HoloLens both have keyboards provided by the OS), and have a custom keyboard as a fallback for systems that don't provide one of their own!

If you have any sort of timeline you'd want to see this functionality in, let me know! I'm happy to factor that into my priorities :) I also added this as an issue at #110

**Reply by TopperDEL** [reply]
Thanks for your quick reply! That feature is definitely quite important for me as my app will be a productivity-based-one and the user needs to enter text.

Your framework is really awesome and I like the way to have "only code" and not all that clutter around like e.g. Unity has. I can start immediately. On the other side the MRTK from MS has plenty of great controls and helpers (like the solvers) that I would miss using StereoKid and I don't really get how much effort it would be to implement those - or use existing features from openXR or whatever framework might help here.
But I guess I'll take StereoKit for my app just because I like how it is structured. Would mean I would very often come back to you and ask for the next feature. :)

**Reply by maluoi** [reply]
Yeah, there's some great stuff in the MRTK, and it'll definitely take me a while to catch up completely! But please feel free to ask for features, I try to prioritize the features that people ask for. Also, I'm super glad to hear you like how StereoKit works! :)

Just note that while there are occasionally other contributors, it's mostly me doing the heavy lifting around here! I'm currently at the tail end of wrapping up v0.3 (a big cross platform overhaul that has taken almost 7 months!), after which I'll be on a _much_ faster cadence of updates. I'll make sure the keyboard stuff is in v0.3.1!

**Reply by TopperDEL** [reply]
That's great to hear! Looking forward to the new features and thank you for being open about new features! It's great to hear that the keyboard-stuff might be added, soon. Keep up the good work!

**Reply by TopperDEL** [reply]
My first version includes speech-to-text currently and it works quite good. For the moment it is enough. But entering text via virtual keyboard is definitely a requirement in the long run.

**Comment by paulmelis** [comment]
Somewhat related, is there a way to get keyboard input in XR when running on a PC with attached XR headset (e.g. an XR-3)? I noticed the StereoKitTest sample Text Input talks about a "soft keyboard" and I'm not sure if that's what I mean. But even when there's a physical keyboard attached to the system running an SK app it would be useful to have an in-XR virtual keyboard that can be used. I would expect most XR usage to be standing anyway, so having go use a physical keyboard is not really ideal in that situation.

Edit: just found `Platform.ForceFallbackKeyboard = true;`
